backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearchResultExtractorImpl.java

```diff
@@ -88,23 +101,57 @@
 		return extractedData;
 	}
 
-	private void addChildrenContributions(IndexSearcher indexSearcher) throws IOException {
-		Document document = storedFieldVisitor.getDocument();
-		String id = document.getField( LuceneFields.idFieldName() ).stringValue();
-		ScoreDoc[] children = findChildren( indexSearcher, id );
-		for ( ScoreDoc childHit : children ) {
-			indexSearcher.doc( childHit.doc , storedFieldVisitor );
+	private Map<Integer, Set<Integer>> fetchNestedDocs(IndexSearcher indexSearcher, ScoreDoc[] scoreDocs) throws IOException {
+		// if the projection does not need any nested object skip their fetching
+		if ( !rootProjection.involveNestedFields() ) {
+			return new HashMap<>();
+		}
+
+		Map<String, Integer> parentIds = new HashMap<>();
+		for ( ScoreDoc hit : scoreDocs ) {
+			Document doc = indexSearcher.doc( hit.doc, ID_FIELD_SET );
+			String parentId = doc.getField( LuceneFields.idFieldName() ).stringValue();
+			if ( parentId == null ) {
+				continue;
+			}
+			parentIds.put( parentId, hit.doc );
+		}
+
+		Map<String, Set<Integer>> stringSetMap = fetchChildren( indexSearcher, parentIds.keySet() );
+		HashMap<Integer, Set<Integer>> result = new HashMap<>();
+		for ( Map.Entry<String, Set<Integer>> entry : stringSetMap.entrySet() ) {
+			result.put( parentIds.get( entry.getKey() ), entry.getValue() );
 		}
+		return result;
 	}
 
-	public ScoreDoc[] findChildren(IndexSearcher indexSearcher, String id) {
-		TermQuery query = new TermQuery( new Term( LuceneFields.rootIdFieldName(), id ) );
+	private Map<String, Set<Integer>> fetchChildren(IndexSearcher indexSearcher, Set<String> parentIds) {
+		BooleanQuery.Builder builder = new BooleanQuery.Builder();
+		for ( String parentId : parentIds ) {
+			TermQuery query = new TermQuery( new Term( LuceneFields.rootIdFieldName(), parentId ) );
+			builder.add( query, BooleanClause.Occur.SHOULD );
+		}
+		builder.setMinimumNumberShouldMatch( 1 );
+
+		BooleanQuery booleanQuery = builder.build();
+		Map<String, Set<Integer>> result = new HashMap<>();
 
 		try {
-			return indexSearcher.search( query, MAX_CHILDREN ).scoreDocs;
+			TopDocs topDocs = indexSearcher.search( booleanQuery, MAX_CHILDREN );
```

Comment:
You don't really need topDocs here, which involves sorting by relevance. `indexSearcher.search( booleanQuery, <some collector>`, where the collector only collects the document IDs in a `DocIdSet`, would probably be better.